File: backend/app/schemas/project.py
from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import Field, field_validator

from app.schemas.base import CamelModel


class ProjectStatus(str, Enum):
    active = "active"
    inactive = "inactive"


class ProjectDraftPayload(CamelModel):
    name: str = Field(min_length=1, max_length=160)
    description: str | None = Field(default=None, max_length=2000)
    task_employee_limit: int | None = Field(default=None, ge=1, le=2_147_483_647)
    status: ProjectStatus = ProjectStatus.active

    @field_validator("name")
    @classmethod
    def validate_non_empty_name(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("Value must not be empty.")
        return value

    @field_validator("description")
    @classmethod
    def validate_optional_description(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = value.strip()
        if not value:
            raise ValueError("Value must not be empty.")
        return value


class ProjectResponse(CamelModel):
    id: str
    name: str
    description: str | None = None
    task_employee_limit: int
    status: ProjectStatus
    created_at: datetime
    updated_at: datetime


class ProjectMemberPayload(CamelModel):
    employee_id: str = Field(min_length=1)


class ProjectMemberSummary(CamelModel):
    employee_id: str
    project_id: str
    employee_name: str
    created_at: datetime

File: frontend/lib/contracts/project.dart
import 'package:touchin_flutter/contracts/contract_parsing.dart';

const int projectNameMaxLength = 160;
const int projectDescriptionMaxLength = 2000;

enum ProjectStatus { active, inactive }

ProjectStatus projectStatusFromApi(String value) {
  return switch (value) {
    'active' => ProjectStatus.active,
    'inactive' => ProjectStatus.inactive,
    _ => throw ContractParsingException(
        'Unsupported project status value: $value',
      ),
  };
}

String projectStatusToApi(ProjectStatus value) {
  return switch (value) {
    ProjectStatus.active => 'active',
    ProjectStatus.inactive => 'inactive',
  };
}

class ProjectSummary {
  const ProjectSummary({
    required this.id,
    required this.name,
    required this.description,
    required this.taskEmployeeLimit,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
  });

  final String id;
  final String name;
  final String? description;
  final int taskEmployeeLimit;
  final ProjectStatus status;
  final DateTime createdAt;
  final DateTime updatedAt;

  factory ProjectSummary.fromJson(JsonMap json) {
    return ProjectSummary(
      id: requireString(json, 'id'),
      name: requireString(json, 'name'),
      description: optionalString(json, 'description'),
      taskEmployeeLimit: requireInt(json, 'taskEmployeeLimit'),
      status: projectStatusFromApi(requireString(json, 'status')),
      createdAt: requireDateTime(json, 'createdAt'),
      updatedAt: requireDateTime(json, 'updatedAt'),
    );
  }
}


class ProjectMemberSummary {
  const ProjectMemberSummary({
    required this.employeeId,
    required this.projectId,
    required this.employeeName,
    required this.createdAt,
  });

  final String employeeId;
  final String projectId;
  final String employeeName;
  final DateTime createdAt;

  factory ProjectMemberSummary.fromJson(JsonMap json) {
    return ProjectMemberSummary(
      employeeId: requireString(json, 'employeeId'),
      projectId: requireString(json, 'projectId'),
      employeeName: requireString(json, 'employeeName'),
      createdAt: requireDateTime(json, 'createdAt'),
    );
  }
}

class ProjectDraft {
  const ProjectDraft({
    required this.name,
    required this.description,
    required this.taskEmployeeLimit,
    this.status = ProjectStatus.active,
  });

  final String name;
  final String description;
  final int taskEmployeeLimit;
  final ProjectStatus status;

  factory ProjectDraft.fromProject(ProjectSummary project) {
    return ProjectDraft(
      name: project.name,
      description: project.description ?? '',
      taskEmployeeLimit: project.taskEmployeeLimit,
      status: project.status,
    );
  }

  JsonMap toApiJson() {
    return <String, dynamic>{
      'name': name.trim(),
      'description': description.trim(),
      'taskEmployeeLimit': taskEmployeeLimit,
      'status': projectStatusToApi(status),
    };
  }
}
